src/metabase/events/notifications.clj

```diff
@@ -0,0 +1,87 @@
+(ns metabase.events.notifications
+  (:require [clojure.core.async :as async]
+            [clojure.tools.logging :as log]
+            [korma.core :as k]
+            [metabase.config :as config]
+            [metabase.email.messages :as messages]
+            [metabase.events :as events]
+            (metabase.models [card :refer [Card]]
+                             [dashboard :refer [Dashboard]]
+                             [dashboard-card :refer [DashboardCard]]
+                             [dependency :refer [Dependency]]
+                             [pulse :refer [Pulse]]
+                             [pulse-card :refer [PulseCard]]
+                             [user :as user])
+            [metabase.util.urls :as urls]))
+
+
+(def notifications-topics
+  "The `Set` of event topics which are subscribed to for use in notifications tracking."
+  #{:segment-update})
+
+(def ^:private notifications-channel
+  "Channel for receiving event notifications we want to subscribe to for notifications events."
+  (async/chan))
+
+
+;;; ## ---------------------------------------- EVENT PROCESSING ----------------------------------------
+
+
+(defn- send-user-segment-notification [segment updated-by user-id card-deps]
+  (let [dash-deps  (when (seq card-deps)
+                     (k/select Dashboard
+                       (k/join DashboardCard (= :report_dashboardcard.dashboard_id :id))
+                       (k/fields :id :name)
+                       (k/where {:report_dashboardcard.card_id [in (mapv :id card-deps)]})))
+        pulse-deps (when (seq card-deps)
+                     (k/select Pulse
+                       (k/join PulseCard (= :pulse_card.pulse_id :id))
+                       (k/fields :id :name)
+                       (k/where {:pulse_card.card_id [in (mapv :id card-deps)]})))
+        send-to    (user/retrieve-user user-id)]
+    (messages/send-notification-email
+      (:email send-to)
+      (:name segment)
+      (:common_name updated-by)
+      (:email updated-by)
+      (mapv #(assoc % :url (urls/dashboard-url (:id %))) dash-deps)
+      (mapv #(assoc % :url (urls/question-url (:id %))) card-deps)
+      (mapv #(assoc % :url (urls/pulse-url (:id %))) pulse-deps)
+      (:revision_message segment))))
```

Comment:
yeah, that stuff was still a work in progress, but i agree it's ugly :cry:

i'll clean it up and make a more sensible method signature out of it.